Multiline textbox fix

RedBandana opened the pull request:
Resolves #278 

The Multiline Textbox was not working like it was supposed to do. It was like a label. With this code, it has all the basic features to expect from a text editor. However, there are no input history, so you will not be able to use Ctr+Z.

**Features**
Basic features to expect from a text editor including:
Keyboard shortcuts: Ctr+C, Ctr+V, Ctr+A, Ctr+X
Horizontal and vertical scrollbar

**Known issues**
Vertical scrollbar doesn't scroll precisely using keyboard directional input if the text has many lines.
The caret (thin black bar) doesn't react precisely to the mouse actions.

lodicolo commented:
@RedBandana So you are aware that this clipboard is only within the engine, and does not allow for clipboard access to the operating system, correct?

Can you please provide video (gif/mp4/etc) demonstration of how the multiline textbox works with the changes you made?

Reference an issue # at the top of this PR (e.g. "Resolves #9999") (create a bug if one does not already exist).

RedBandana commented:
> @RedBandana So you are aware that this clipboard is only within the engine, and does not allow for clipboard access to the operating system, correct?
Yeah I was aware of that

> Can you please provide video (gif/mp4/etc) demonstration of how the multiline textbox works with the changes you made?
> Reference an issue # at the top of this PR (e.g. "Resolves #9999") (create a bug if one does not already exist).
Ok, will do that.

RedBandana commented:
While doing the video, I realized there were still some issues that need to be fixed in order to have a good product. I guess I will fix those and update the pull request.

**Issues found**
Client crashes when trying to paste a text when nothing has been copied.
Some json properties are not taken in consideration like the "TextColor" one.

RedBandana commented:
Actually, the current multiline text box is not usable as we cannot type multiple lines or create a json with it [(preview here)](https://s3.us-east-2.amazonaws.com/ascensiongamedev/filehost/ffa6c928a3f248a4cfbf03d5e2d573d5.gif). This code is here to help on that. Basically, it generates a label for each line.

So there are two new parameters when instancing a multiline textbox: a string for the name of the json file, and a boolean to allow the write permission.
 `mMultiTextbox = new MultilineTextBox(mTextboxBg, "TextBoxText", canEdit);`

The image below represents a json file generated by the code above. Values inside the green boxes are the ones shared with the text (more precisely the labels). The other ones are not taken in consideration or apply to the text container only.
![](https://ascensiongamedev.com/resources/filehost/c9bc110b9818aed2b5c38c57ab41aad5.png)

You can end up with something like this when using several Gwen Control classes.
![](https://ascensiongamedev.com/resources/filehost/392e0077839b255eed2aa7277588f1de.gif)

RedBandana commented on Intersect.Client.Framework/Gwen/Control/MultilineTextBox.cs:
```diff
@@ -58,16 +59,14 @@ public MultilineTextBox(Base parent) : base(parent)
-            AddAccelerator("Ctrl + X", OnCut);
-            AddAccelerator("Ctrl + V", OnPaste);
-            AddAccelerator("Ctrl + A", OnSelectAll);
+            AddAccelerator("CTRL+C", OnCopy);
```
It was not working with spaces.

RedBandana commented on Intersect.Client.Framework/Gwen/Control/MultilineTextBox.cs:
```diff
@@ -1116,10 +1238,10 @@ public override void SetText(string str, bool doEvents = true)
-            {
-                mText.String = Text;
-            }
+            //if (mText != null)
```
To be honest I don't remember. Probably an oversight. Will check that out when I will edit the code.

RedBandana commented on Intersect.Client.Framework/Gwen/Control/MultilineTextBox.cs:
```diff
@@ -190,13 +189,9 @@ public override string Text
             {
                 var ret = "";
-                for (var i = 0; i < TotalLines; i++)
+                foreach (var item in mTextLines)
```
Indeed, will edit that.

RedBandana replied:
Finally it's because labels are used in mTextLines, so the easiest working way to proceed I found was to do that.

RedBandana commented on Intersect.Client.Framework/Gwen/Control/MultilineTextBox.cs:
```diff
@@ -970,6 +1063,20 @@ protected override void OnMouseClickedLeft(int x, int y, bool down, bool automat
             RefreshCursorBounds();
         }
 
+        private Label GenerateLabel(Label label, string text = "")
```
Oh ok I see.

RedBandana commented on Intersect.Client.Framework/Gwen/Control/MultilineTextBox.cs:
```diff
@@ -506,6 +552,23 @@ protected override bool OnKeyReturn(bool down)
+        private void RemoveLabel(int index)
+        {
+            mTextLines[index].Dispose(); // Have to dispose so the text disappears
+            //mTextLines[index] = null; // Have to set the value to null in order for the vertical scroll bar to update.
```
Probably an oversight when debugging. Will change that on next update.

RedBandana commented:
Thanks for the comments @lodicolo. I will apply all requests and suggestions when I will update the branch.

RedBandana commented:
Hello @lodicolo , I've updated the code of the branch based on your comments and fixed some issues I discovered.

**Known issues:**
Vertical scrollbar doesn't scroll precisely using keyboard directional input if the text has many lines.
The caret (thin black bar) doesn't react precisely to the mouse actions.
Adding new lines by pressing Enter is somehow slow
